src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,221 @@ template<class Arc> class DeterminizerStar {
 
       if (max_states_ > 0 && counter++ > max_states_) {
         std::cerr << "Determinization aborted since looped more than "
                   << max_states_ << " times during epsilon closure.\n";
-        throw std::runtime_error("looped more than max-states times in determinization");
-      }
-      for (ArcIterator<Fst<Arc> > aiter(*ifst_, elem.state);
-           !aiter.Done(); aiter.Next()) {
-        const Arc &arc = aiter.Value();
-        if (sorted && arc.ilabel > 0) {
-          break;
-          // Break from the loop: due to sorting there will be no
-          // more transitions with epsilons as input labels.
-        }
-        if (arc.ilabel == 0) {  // Epsilon transition.
-          Element next_elem;
-          next_elem.state = arc.nextstate;
-          next_elem.weight = Times(elem.weight, arc.weight);
-          // now must append strings
-          if (arc.olabel == 0)
-            next_elem.string = elem.string;
-          else {
-            vector<Label> seq;
-            repository_.SeqOfId(elem.string, &seq);
-            if (arc.olabel != 0)
-              seq.push_back(arc.olabel);
-            next_elem.string = repository_.IdOfSeq(seq);
-          }
-          typename std::map<InputStateId, Element>::iterator
-              iter = cur_subset.find(next_elem.state);
-          if (iter == cur_subset.end()) {
-            // was no such StateId: insert and add to queue.
-            cur_subset[next_elem.state] = next_elem;
-            queue.push_back(next_elem);
-          } else {  // one is already there.  Add weights.
-            if (iter->second.string != next_elem.string) {
-              std::cerr << "DeterminizerStar: FST was not functional -> not determinizable\n";
-              { // Print some debugging information.  Can be helpful to debug
-                // the inputs when FSTs are mysteriously non-functional.
-                vector<Label> tmp_seq;
-                repository_.SeqOfId(iter->second.string, &tmp_seq);
-                std::cerr << "First string: ";
-                for (size_t i = 0; i < tmp_seq.size(); i++) std::cerr << tmp_seq[i] << " ";
-                std::cerr << "\nSecond string: ";
-                repository_.SeqOfId(next_elem.string, &tmp_seq);
-                for (size_t i = 0; i < tmp_seq.size(); i++) std::cerr << tmp_seq[i] << " ";
-                std::cerr << "\n";
-              }
-              throw std::runtime_error("Non-functional FST: cannot determinize.\n");
-            }
-            Weight weight = Plus(iter->second.weight, next_elem.weight);
-            if (! ApproxEqual(weight, iter->second.weight, delta_)) {  // add extra part of weight to queue.
-              iter->second.weight = weight; // Update weight in map.
-              queue.push_back(next_elem);
-              replaced_elems = true;
-            }
-          }
-        }
+        throw std::runtime_error("looped more than max-states times"
+           " in determinization");
       }
+
```

Comment:
You need to be a bit careful about iterator invalidation here.  I think you are passing to ExpandOneElement a reference to 'elem' that could be invalidated inside ExpandOneElement.
Also, in general, you should avoid an iterator being invalidated while it is still in scope, or you should at least have a comment to explain what's going on.